File: 3DEngine/Delaunay/Line.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Delaunay {
    public class Line {
        public Point3d Start, End;
        public Line(Point3d start, Point3d end) {
            this.Start = start;
            this.End = end;
        }

        // Swap start and end points
        public void Reverse() {
            Point3d tmp = this.Start;
            this.Start = this.End;
            this.End = tmp;
        }

        public override bool Equals(object obj) {
            if(obj is Line) {
                Line l = (Line)obj;
                return ((this.Start.Equals(l.Start)) && (this.End.Equals(l.End)))
                    || ((this.Start.Equals(l.End)) && (this.End.Equals(l.Start)));
            } else {
                return false;
            }
        }

        public Point3d[] Vertices() {
            return new Point3d[] { Start, End };
        }

        public override int GetHashCode() {
            return base.GetHashCode();
        }

        public override string ToString() {
            return string.Format("{0} -> {1}", Start.ToString(), End.ToString());
        }

        public static Line operator +(Line v1, Line v2) {
            return new Line(v1.Start + v2.Start, v1.End + v2.End);
        }

        public static Line operator +(Line v1, Point3d p) {
            return new Line(v1.Start + p, v1.End + p);
        }

        public static Line operator -(Line v1, Line v2) {
            return new Line(v1.Start - v2.Start, v1.End - v2.End);
        }

        public static Line operator -(Line v1, Point3d p) {
            return new Line(v1.Start - p, v1.End - p);
        }

        public static Line operator *(Line v1, double scalar) {
            return new Line(v1.Start * scalar, v1.End * scalar);
        }

        public static Line operator *(double scalar, Line v1) {
            return v1 * scalar;
        }
    }
}

File: 3DEngine/Claases/Delaunay/Line.vb
Namespace Delaunay2
    Public Class Line
        Implements IEquatable(Of Line)

        Public Start As Point3d
        Public [End] As Point3d

        Public Sub New(start As Point3d, [end] As Point3d)
            Me.Start = start
            Me.End = [end]
        End Sub

        Public Sub Reverse()
            Dim tmp As Point3d = Start
            Start = [End]
            [End] = tmp
        End Sub

        Public Function Vertices() As Point3d()
            Return {Start, [End]}
        End Function

        Public Shared Operator =(l1 As Line, l2 As Line) As Boolean
            Return (l1.Start = l2.Start AndAlso l1.End = l2.End) OrElse
                   (l1.Start = l2.End AndAlso l1.End = l2.Start)
        End Operator

        Public Shared Operator <>(l1 As Line, l2 As Line) As Boolean
            Return Not (l1 = l2)
        End Operator

        Public Shared Operator +(l1 As Line, l2 As Line) As Line
            Return New Line(l1.Start + l2.Start, l1.End + l2.End)
        End Operator

        Public Shared Operator +(l1 As Line, p As Point3d) As Line
            Return New Line(l1.Start + p, l1.End + p)
        End Operator

        Public Shared Operator -(l1 As Line, l2 As Line) As Line
            Return New Line(l1.Start - l2.Start, l1.End - l2.End)
        End Operator

        Public Shared Operator -(l1 As Line, p As Point3d) As Line
            Return New Line(l1.Start - p, l1.End - p)
        End Operator

        Public Shared Operator *(l1 As Line, scalar As Double) As Line
            Return New Line(l1.Start * scalar, l1.End * scalar)
        End Operator

        Public Shared Operator *(scalar As Double, l1 As Line) As Line
            Return l1 * scalar
        End Operator

        Public Shadows Function Equals(other As Line) As Boolean Implements IEquatable(Of Line).Equals
            Return Me = other
        End Function

        Public Overrides Function ToString() As String
            Return $"{Start} -> {[End]}"
        End Function
    End Class
End Namespace
